use table_widget

superstar54 opened the pull request:
Use the new [Table wdiget](https://github.com/superstar54/table-widget) which supports editing, filtering features.

Supports:
- to edit `label` and `description` of the job
- custom columns by show/hide columns
- a quick search bar to find jobs by any visible field.
- add `exit_status` and `exit_message`

https://github.com/user-attachments/assets/09506462-5617-475a-9ce7-9773d8900db0

edan-bainglass commented on calculation_history.ipynb:
```diff
@@ -50,24 +50,34 @@
     "        }\n",
     "    </style>\n",
-    "    <div class=\"page-title\">Calculation history</div>\n",
+    "    <div class=\"page-title\">Calculation History</div>\n",
```
Leave all titles in sentence case

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -1,24 +1,61 @@
+from aiidalab_qe.common.widgets import LoadingWidget
 
-state_icons = {
+STATE_ICONS = {
     "running": "⏳",
     "finished": "✅",
     "excepted": "⚠️",
     "killed": "❌",
 }
```
I use mostly the same icons in the process tree. Best we agree on an icon set and place it in the common module.

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -1,24 +1,61 @@
     "killed": "❌",
 }
 
+COLUMNS = {
+    "ID": {"headerName": "ID 🔗", "dataType": "link", "editable": False},
+    "Creation time absolute": {
+        "headerName": "Creation Time ⏰ (absolute)",
+        "type": "date",
+        "width": 150,
+        "editable": False,
+    },
+    "Creation time relative": {
+        "headerName": "Creation Time ⏰ (relative)",
+        "width": 150,
+        "editable": False,
+    },
+    "Structure": {"headerName": "Structure", "editable": False},
+    "State": {"headerName": "State 🟢", "editable": False},
+    "Label": {"headerName": "Label", "width": 300, "editable": True},
+    "Description": {
+        "headerName": "Description",
+        "width": 300,
+        "editable": True,
+        "hide": True,
+    },
+    "Relax_type": {"headerName": "Relax_type", "editable": False, "hide": True},
+    "Delete": {"headerName": "Delete", "dataType": "link", "editable": False},
+    "Download": {"headerName": "Download", "dataType": "link", "editable": False},
+    "UUID": {"headerName": "UUID", "editable": False, "hide": True},
+}
+
```
Sentence case everywhere please 🙏 Also, `Relax_type` should be `Relax type`. No underscores please 🙏

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -1,24 +1,61 @@
+class CalculationHistory:
     def __init__(self):
-        pass
+        self.main = ipw.VBox(children=[LoadingWidget("Loading the table...")])
+
+        self.table = TableWidget(style={"margin-top": "20px"})
 
-    def setup_table(self):
+        def on_row_update(change):
+            node = load_node(change["new"]["UUID"])
+            node.label = change["new"]["Label"]
+            node.description = change["new"]["Description"]
+
+        self.table.observe(on_row_update, "updatedRow")
```
If down the road someone decided to integrate this class into their app (within the app), this implementation will cause the table to load on app start, even if the table is three tabs deep and not visited often. In other words, I strongly recommend you restructure with lazy loading in mind (i.e., at least a `render` method to call when needed.

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -1,24 +1,61 @@
+        self.table.observe(on_row_update, "updatedRow")
+
+    def load_table(self):
         self.df = self.load_data()
-        self.table = ipw.HTML()
         self.setup_widgets()
 
     def load_data(self):
```
Ignore my comment above regarding lazy loading if indeed these load methods are only called "just in time" and are the bulk "cost" of the widget. Otherwise, my comment above is still valid.

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -183,57 +205,67 @@ def setup_widgets(self):
+            children=[
+                ipw.HTML("<h4>Filters:</h4>"),
                 ipw.VBox(
                     [
```
```suggestion
                    children=[
```

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -183,57 +205,67 @@ def setup_widgets(self):
+                        self.job_state_dropdown,
+                        self.time_box,
+                        ipw.VBox(
+                            [self.properties_filter_description, self.properties_box],
```
```suggestion
                            children=[
                                self.properties_filter_description, 
                                self.properties_box,
                            ],
```

edan-bainglass commented on src/aiidalab_qe/app/utils/search_jobs.py:
```diff
@@ -183,57 +205,67 @@ def setup_widgets(self):
+                        ipw.VBox(
+                            [self.properties_filter_description, self.properties_box],
+                            layout=ipw.Layout(
+                                border="1px solid #ddd", padding="5px", margin="5px"
```
```suggestion
                                border="1px solid #ddd", 
                                padding="5px", 
                                margin="5px",
```

I promise I'm not being picky 😅 This line by line format makes git maintenance easier.

edan-bainglass reviewed (changes requested):
Very nice @superstar54! Please see my comments.

AndresOrtegaGuerrero commented:
Nice @superstar54 , I was wondering , can #1045 be address in the PR ?

superstar54 commented:
> I was wondering , can #1045 be address in the PR ?

Yes, thanks for pointing out this. I added this feature as shown below:

![Screenshot from 2025-01-09 17-04-19](https://github.com/user-attachments/assets/70ec6d76-1490-46b3-9af7-44faaf636dd6)

superstar54 commented:
Hi @edan-bainglass , I resolved all your comments except the icons. We can unify the icons later.

edan-bainglass commented:
> > I was wondering , can #1045 be address in the PR ?
> 
> Yes, thanks for pointing out this. I added this feature as shown below:

Remember to close #1045 if done

edan-bainglass commented:
![Screenshot 2025-01-09 184941](https://github.com/user-attachments/assets/3b6d3363-69ea-41b6-8c53-e2c3193d9010)

1. Header is too much. Keep it as it was before, but sentence case
2. Page guide button is perfect!
3. Remove the border from everything. It's not adding anything. If you disagree, make sure to add padding to anything that has a border, otherwise, it's looks too cramped.

No comments regarding the code. Thank you for applying the previous suggestions. Once these points are addressed, you're good to go.

superstar54 commented:
> I have plenty of jobs. Why none showing up?

How do you install the `table_widget` from the repo or use pip? If from repo, you also need to run `npm install & npm run build`


Hi @edan-bainglass , I have resolved all other issues.

edan-bainglass commented:
> > I have plenty of jobs. Why none showing up?
> 
> How do you install the `table_widget` from the repo or use pip? If from repo, you also need to run `npm install & npm run build`

I'm installing from pypi. I've upgraded to 0.0.2. Still no table showing!

> Hi @edan-bainglass , I have resolved all other issues.

![Screenshot 2025-01-10 074243](https://github.com/user-attachments/assets/eb277b94-e1de-4487-9646-7c9f64d20f2e)

Simple header (as we do for all notebook except the app). `h1` tag, no CSS class. Maybe add some bottom margin if you want. 
We can discuss next week if we want all notebook to carry the banner, but that should be for the next release (or minor release). 

Also, I really think you should remove the borders. I also removed the margin from the table component. It is much cleaner without it.

edan-bainglass commented:
> ![Screenshot 2025-01-10 074243](https://private-user-images.githubusercontent.com/45081142/401850602-eb277b94-e1de-4487-9646-7c9f64d20f2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0OTE4OTIsIm5iZiI6MTczNjQ5MTU5MiwicGF0aCI6Ii80NTA4MTE0Mi80MDE4NTA2MDItZWIyNzdiOTQtZTFkZS00NDg3LTk2NDYtN2M5ZjY0ZDIwZjJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDA2NDYzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwMTYyNWViOTAzNTdkNzI2ZDZlOGRjNjU3YmZmMzIzZWE2ZDRhNGNlYWQyMDQ5MTI0NzJkYTBjYzM0ZWVlZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hEULNQRY2OyxJ3BdWYV_CPUYRsB7oKgXxYBUt5jFPDU)
> 
> Simple header (as we do for all notebook except the app). `h1` tag, no CSS class. Maybe add some bottom margin if you want. We can discuss next week if we want all notebook to carry the banner, but that should be for the next release (or minor release).
> 
> Also, I really think you should remove the borders. I also removed the margin from the table component. It is much cleaner without it.

Alternatively, if we do want borders

![Screenshot 2025-01-10 075616](https://github.com/user-attachments/assets/7e56a2b8-01da-4e33-8a97-3065962d04d2)

### Styles
- Heading: `h1` with 15px bottom margin
- boxes: 
  - border: 1px solid lightgray
  - padding: 0.5em
- properties filter box
  - same as above, but also margin: 5px

superstar54 commented:
> We can discuss next week if we want all notebook to carry the banner, but that should be for the next release (or minor release).

agree, we should consider to uniform the page header and footer for all notebooks.

> I'm installing from pypi. I've upgraded to 0.0.2. Still no table showing!

Any error in the browser console?


Thanks for for the style suggestion, I have implemented the styles.

edan-bainglass commented:
> > We can discuss next week if we want all notebook to carry the banner, but that should be for the next release (or minor release).
> 
> agree, we should consider to uniform the page header and footer for all notebooks.
> 
> > I'm installing from pypi. I've upgraded to 0.0.2. Still no table showing!
> 
> Any error in the browser console?
> 
> Thanks for for the style suggestion, I have implemented the styles.

Not sure why, but it works now. I can see my jobs. Styles look great! However, I am getting an error.

Filtered by PDOS, then by running job state, and got

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/opt/conda/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in _handle_msg(self, msg)
    762                 if 'buffer_paths' in data:
    763                     _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 764                 self.set_state(state)
    765 
    766         # Handle a state request.

/opt/conda/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in set_state(self, sync_data)
    631                     from_json = self.trait_metadata(name, 'from_json',
    632                                                     self._trait_from_json)
--> 633                     self.set_trait(name, from_json(sync_data[name], self))
    634 
    635     def send(self, content, buffers=None):

/opt/conda/lib/python3.9/contextlib.py in __exit__(self, typ, value, traceback)
    124         if typ is None:
    125             try:
--> 126                 next(self.gen)
    127             except StopIteration:
    128                 return False

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in hold_trait_notifications(self)
   1500                 for changes in cache.values():
   1501                     for change in changes:
-> 1502                         self.notify_change(change)
   1503 
   1504     def _notify_trait(self, name, old_value, new_value):

/opt/conda/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in notify_change(self, change)
    692                 # Send new state to front-end
    693                 self.send_state(key=name)
--> 694         super(Widget, self).notify_change(change)
    695 
    696     def __repr__(self):

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in notify_change(self, change)
   1515     def notify_change(self, change):
   1516         """Notify observers of a change event"""
-> 1517         return self._notify_observers(change)
   1518 
   1519     def _notify_observers(self, event):

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in _notify_observers(self, event)
   1562                 c = getattr(self, c.name)
   1563 
-> 1564             c(event)
   1565 
   1566     def _add_notifiers(self, handler, name, type):

/opt/conda/lib/python3.9/site-packages/ipywidgets/widgets/widget_selection.py in _propagate_index(self, change)
    233             self.label = label
    234         if self.value is not value:
--> 235             self.value = value
    236 
    237     @validate('value')

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in __set__(self, obj, value)
    730             raise TraitError('The "%s" trait is read-only.' % self.name)
    731         else:
--> 732             self.set(obj, value)
    733 
    734     def _validate(self, obj, value):

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in set(self, obj, value)
    719             # we explicitly compare silent to True just in case the equality
    720             # comparison above returns something other than True/False
--> 721             obj._notify_trait(self.name, old_value, new_value)
    722 
    723     def __set__(self, obj, value):

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in _notify_trait(self, name, old_value, new_value)
   1503 
   1504     def _notify_trait(self, name, old_value, new_value):
-> 1505         self.notify_change(
   1506             Bunch(
   1507                 name=name,

/opt/conda/lib/python3.9/site-packages/ipywidgets/widgets/widget.py in notify_change(self, change)
    692                 # Send new state to front-end
    693                 self.send_state(key=name)
--> 694         super(Widget, self).notify_change(change)
    695 
    696     def __repr__(self):

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in notify_change(self, change)
   1515     def notify_change(self, change):
   1516         """Notify observers of a change event"""
-> 1517         return self._notify_observers(change)
   1518 
   1519     def _notify_observers(self, event):

~/.local/lib/python3.9/site-packages/traitlets/traitlets.py in _notify_observers(self, event)
   1562                 c = getattr(self, c.name)
   1563 
-> 1564             c(event)
   1565 
   1566     def _add_notifiers(self, handler, name, type):

~/apps/quantum-espresso/src/aiidalab_qe/app/utils/search_jobs.py in apply_filters(self, _)
    335                 & (filtered_df["ctime"] <= end_time)
    336             ]
--> 337         self.update_table_value(filtered_df)
    338 
    339     def update_table_visibility(self, _):

~/apps/quantum-espresso/src/aiidalab_qe/app/utils/search_jobs.py in update_table_value(self, display_df)
    291         # Adjust the ID column based on the toggle state
    292         if self.toggle_id_format.value == "PK":
--> 293             display_df = display_df.rename(columns={"PK_with_link": "ID"}).drop(
    294                 columns=["UUID_with_link"]
    295             )

~/.local/lib/python3.9/site-packages/pandas/core/frame.py in drop(self, labels, axis, index, columns, level, inplace, errors)
   5579                 weight  1.0     0.8
   5580         """
-> 5581         return super().drop(
   5582             labels=labels,
   5583             axis=axis,

~/.local/lib/python3.9/site-packages/pandas/core/generic.py in drop(self, labels, axis, index, columns, level, inplace, errors)
   4786         for axis, labels in axes.items():
   4787             if labels is not None:
-> 4788                 obj = obj._drop_axis(labels, axis, level=level, errors=errors)
   4789 
   4790         if inplace:

~/.local/lib/python3.9/site-packages/pandas/core/generic.py in _drop_axis(self, labels, axis, level, errors, only_slice)
   4828                 new_axis = axis.drop(labels, level=level, errors=errors)
   4829             else:
-> 4830                 new_axis = axis.drop(labels, errors=errors)
   4831             indexer = axis.get_indexer(new_axis)
   4832 

~/.local/lib/python3.9/site-packages/pandas/core/indexes/base.py in drop(self, labels, errors)
   7068         if mask.any():
   7069             if errors != "ignore":
-> 7070                 raise KeyError(f"{labels[mask].tolist()} not found in axis")
   7071             indexer = indexer[~mask]
   7072         return self.delete(indexer)

KeyError: "['UUID_with_link'] not found in axis"
```

superstar54 commented:
> Filtered by PDOS, then by running job state, and got

This is because no result is found, and the pd.dataframe is empty. I fixed it, and it can handle empty data. Besides, I also dropped the usage of `pandas` because I remember @danielhollas mentioned that he wanted to drop `pandas` in the AWB. The performance seems not affected.

edan-bainglass reviewed (approved):
Excellent work @superstar54. Thanks! I'll open a PR to use it more widely in the app.
